README.md
# `argtyped`: Command Line Argument Parser, with Types

[![Build Status](https://github.com/huzecong/argtyped/workflows/Build/badge.svg)](https://github.com/huzecong/argtyped/actions?query=workflow%3ABuild+branch%3Amaster)
[![CodeCov](https://codecov.io/gh/huzecong/argtyped/branch/master/graph/badge.svg?token=ELHfYJ2Ydq)](https://codecov.io/gh/huzecong/argtyped)
[![PyPI](https://img.shields.io/pypi/v/argtyped.svg)](https://pypi.org/project/argtyped/)
[![MIT License](https://img.shields.io/badge/license-MIT-blue.svg)](https://github.com/huzecong/argtyped/blob/master/LICENSE)

`argtyped` is an command line argument parser with that relies on type annotations. It is built on
[`argparse`](https://docs.python.org/3/library/argparse.html), the command line argument parser library built into
Python. Compared with `argparse`, this library gives you:

- More concise and intuitive syntax, less boilerplate code.
- Type checking and IDE auto-completion for command line arguments.
- A drop-in replacement for `argparse` in most cases.

Since v0.4.0, `argtyped` also supports parsing arguments defined with an [attrs](https://attrs.org/)-class. See
[Attrs Support](#attrs-support-new) for more details.  


## Installation

Install stable release from [PyPI](https://pypi.org/project/argtyped/):
```bash
pip install argtyped
```

Or, install the latest commit from GitHub:
```bash
pip install git+https://github.com/huzecong/argtyped.git
```

## Usage

With `argtyped`, you can define command line arguments in a syntax similar to
[`typing.NamedTuple`](https://docs.python.org/3/library/typing.html#typing.NamedTuple). The syntax is intuitive and can
be illustrated with an example:
```python
from typing import Optional
from typing_extensions import Literal  # or directly import from `typing` in Python 3.8+

from argtyped import Arguments, Switch
from argtyped import Enum, auto

class LoggingLevels(Enum):
    Debug = auto()
    Info = auto()
    Warning = auto()
    Error = auto()
    Critical = auto()

class MyArguments(Arguments):
    model_name: str         # required argument of `str` type
    hidden_size: int = 512  # `int` argument with default value of 512

    activation: Literal["relu", "tanh", "sigmoid"] = "relu"  # argument with limited choices
    logging_level: LoggingLevels = LoggingLevels.Info        # using `Enum` class as choices

    use_dropout: Switch = True  # switch argument, enable with "--use-dropout" and disable with "--no-use-dropout"
    dropout_prob: Optional[float] = 0.5  # optional argument, "--dropout-prob=none" parses into `None`

args = MyArguments()
```

This is equivalent to the following code with Python built-in `argparse`:
```python
import argparse
from enum import Enum

class LoggingLevels(Enum):
    Debug = "debug"
    Info = "info"
    Warning = "warning"
    Error = "error"
    Critical = "critical"

parser = argparse.ArgumentParser()

parser.add_argument("--model-name", type=str, required=True)
parser.add_argument("--hidden-size", type=int, default=512)

parser.add_argument("--activation", choices=["relu", "tanh", "sigmoid"], default="relu")
parser.add_argument("--logging-level", choices=list(LoggingLevels), type=LoggingLevels, default="info")

parser.add_argument("--use-dropout", action="store_true", dest="use_dropout", default=True)
parser.add_argument("--no-use-dropout", action="store_false", dest="use_dropout")
parser.add_argument("--dropout-prob", type=lambda s: None if s.lower() == 'none' else float(s), default=0.5)

args = parser.parse_args()
```

Save the code into a file named `main.py`. Suppose the following arguments are provided:
```bash
python main.py \
    --model-name LSTM \
    --activation sigmoid \
    --logging-level debug \
    --no-use-dropout \
    --dropout-prob none
```
Then the parsed arguments will be equivalent to the following structure returned by `argparse`:
```python
argparse.Namespace(
    model_name="LSTM", hidden_size=512, activation="sigmoid", logging_level="debug",
    use_dropout=False, dropout_prob=None)
```

Arguments can also be pretty-printed:
```
>>> print(args)
<class '__main__.MyArguments'>
╔═════════════════╤══════════════════════════════════╗
║ Arguments       │ Values                           ║
╠═════════════════╪══════════════════════════════════╣
║ model_name      │ 'LSTM'                           ║
║ hidden_size     │ 512                              ║
║ activation      │ 'sigmoid'                        ║
║ logging_level   │ <MyLoggingLevels.Debug: 'debug'> ║
║ use_dropout     │ False                            ║
║ dropout_prob    │ None                             ║
║ label_smoothing │ 0.1                              ║
║ some_true_arg   │ True                             ║
║ some_false_arg  │ False                            ║
╚═════════════════╧══════════════════════════════════╝
```
It is recommended though to use the `args.to_string()` method, which gives you control of the table width.

## Attrs Support (New)

The way we define the arguments is very similar to defining a [dataclass](https://docs.python.org/3/library/dataclasses.html)
or an [attrs](https://attrs.org)-class, so it seems natural to just write an attrs-class, and add parsing support to it.

To use `argtyped` with `attrs`, simply define an attrs-class as usual, and have it subclass `AttrsArguments`. Here's 
the same example above, but implemented with attrs-classes, and with some bells and whistles:
```python
import attr  # note: new style `attrs` syntax is also supported
from argtyped import AttrsArguments

def _convert_logging_level(s: str) -> LoggingLevels:
    # Custom conversion function that takes the raw string value from the command line.
    return LoggingLevels[s.lower()]

@attr.s(auto_attribs=True)
class MyArguments(AttrsArguments):
    model_name: str = attr.ib(metadata={"positional": True})  # positional argument
    # Or: `model_name: str = argtyped.positional_arg()`.
    layer_sizes: List[int] = attr.ib(metadata={"nargs": "+"})  # other metadata are treated as `argparse` options

    activation: Literal["relu", "tanh", "sigmoid"] = "relu"
    logging_level: LoggingLevels = attr.ib(default=LoggingLevels.Info, converter=_convert_logging_level)

    use_dropout: Switch = True
    dropout_prob: Optional[float] = 0.5

    _activation_fn: Callable[[float], float] = attr.ib(init=False)  # `init=False` attributes are not parsed

    @dropout_prob.validator  # validators still work as you would expect
    def _dropout_prob_validator(self, attribute, value):
        if not 0.0 <= value <= 1.0:
            raise ValueError(f"Invalid probability {value}")

    @_activation_fn.default
    def _activation_fn(self):
        return _ACTIVATION_FNS[self.activation]
```

A few things to note here:
- You can define positional arguments by adding `"positional": True` as metadata. If this feels unnatural, you could 
  also use `argtyped.positional_arg()`, which takes the same arguments as `attr.ib`.
- You can pass additional options to `ArgumentParser.add_argument` by listing them as metadata as well. Note that 
  these options take precedence over `argtyped`'s computed arguments, for example, sequence arguments (`List[T]`) by 
  default uses `nargs="*"`, but you could override it with metadata.
- Attributes with custom converters will not be parsed; its converter will be called with the raw string value from
  command line. If the attribute also has a default value, you should make sure that your converter works with both
  strings and the default value.
- `init=False` attributes are not treated as arguments, but they can be useful for storing computed values based on
  arguments.
- The default value logic is the same as normal attrs classes, and thus could be different from non-attrs `argtyped` 
  classes. For example, optional arguments are not considered to have an implicit default of `None`, and no type 
  validation is performed on default values.

Here are the (current) differences between an attrs-based arguments class (`AttrsArguments`) versus the normal arguments
class (`Arguments`):
- `AttrsArguments` supports positional arguments and custom options via metadata.
- `AttrsArguments` handles default values with attrs, so there's no validation of default value types.  This also 
  means that nullable arguments must have an explicit default value of `None`, otherwise it becomes a required 
  argument.
- `AttrsArguments` does not support `underscore=True`.
- `AttrsArguments` does not have `to_dict`, `to_string` methods.
- `AttrsArguments` needs to be called with the factory `parse_args` method to parse, while `Arguments` parses command 
  line arguments on construction.


## Reference

### The `argtyped.Arguments` Class

The `argtyped.Arguments` class is main class of the package, from which you should derive your custom class that holds
arguments. Each argument takes the form of a class attribute, with its type annotation and an optional default value.

When an instance of your custom class is initialized, the command line arguments are parsed from `sys.argv` into values
with your annotated types. You can also provide the list of strings to parse by passing them as the parameter.

The parsed arguments are stored in an object of your custom type. This gives you arguments that can be auto-completed
by the IDE, and type-checked by a static type checker like [mypy](http://mypy-lang.org/).

The following example illustrates the keypoints:
```python
class MyArgs(argtyped.Arguments):
    # name: type [= default_val]
    value: int = 0

args = MyArgs()                    # equivalent to `parser.parse_args()`
args = MyArgs(["--value", "123"])  # equivalent to `parser.parse_args(["--value", "123"])
assert isinstance(args, MyArgs)
```

#### `Arguments.to_dict(self)`

Convert the set of arguments to a dictionary (`OrderedDict`).

#### `Arguments.to_string(self, width: Optional[int] = None, max_width: Optional[int] = None)`

Represent the arguments as a table.
- `width`: Width of the printed table. Defaults to `None`, which fits the table to its contents. An exception is raised
  when the table cannot be drawn with the given width.
- `max_width`: Maximum width of the printed table. Defaults to `None`, meaning no limits. Must be `None` if `width` is
  not `None`.

#### `argtyped.argument_specs`

Return a dictionary mapping argument names to their specifications, represented as the `argtyped.ArgumentSpec` type.
This is useful for programmatically accessing the list of arguments.

### Argument Types

To summarize, whatever works for `argparse` works here. The following types are supported:

- **Built-in types** such as `int`, `float`, `str`.
- **Boolean type** `bool`. Accepted values (case-insensitive) for `True` are: `y`, `yes`, `true`, `ok`; accepted values
  for `False` are: `n`, `no`, `false`.
- **Choice types** `Literal[...]`. A choice argument is essentially an `str` argument with limited
  choice of values. The ellipses can be filled with a tuple of `str`s, or an expression that evaluates to a list of
  `str`s:
  ```python
  from argtyped import Arguments
  from typing_extensions import Literal

  class MyArgs(Arguments):
      foo: Literal["debug", "info", "warning", "error"]  # 4 choices

  # argv: ["--foo=debug"] => foo="debug"
  ```
  This is equivalent to the `choices` keyword in `argparse.add_argument`.
  
  **Note:** The choice type was previously named `Choices`. This is deprecated in favor of the
  [`Literal` type](https://mypy.readthedocs.io/en/stable/literal_types.html) introduced in Python 3.8 and back-ported to
  3.6 and 3.7 in the `typing_extensions` library. `Choices` was removed since version 0.4.0.
- **Enum types** derived from `enum.Enum`. It is recommended to use `argtyped.Enum` which uses the instance names as
  values:
  ```python
  from argtyped import Enum

  class MyEnum(Enum):
      Debug = auto()    # "debug"
      Info = auto()     # "info"
      Warning = auto()  # "warning"
  ```
- **Switch types** `Switch`. `Switch` arguments are like `bool` arguments, but they don't take values. Instead, a switch
  argument `switch` requires `--switch` to enable and `--no-switch` to disable:
  ```python
  from argtyped import Arguments, Switch

  class MyArgs(Arguments):
      switch: Switch = True
      bool_arg: bool = False

  # argv: []                                 => flag=True,  bool_arg=False
  # argv: ["--switch", "--bool-arg=false"]   => flag=True,  bool_arg=False
  # argv: ["--no-switch", "--bool-arg=true"] => flag=False, bool_arg=True
  # argv: ["--switch=false"]                 => WRONG
  # argv: ["--no-bool-arg"]                  => WRONG
  ```
- **List types** `List[T]`, where `T` is any supported type except switch types. List arguments allow passing multiple
  values on the command line following the argument flag, it is equivalent to setting `nargs="*"` in `argparse`.
  
  Although there is no built-in support for other `nargs` settings such as `"+"` (one or more) or `N` (fixed number),
  you can add custom validation logic by overriding the `__init__` method in your `Arguments` subclass.
- **Optional types** `Optional[T]`, where `T` is any supported type except list or switch types. An optional argument

CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [0.4.0]
### Added
- Support for parsing arguments based on `attrs` attributes.

### Changed
- Removed `Choices` in favor of `Literal`.
- Changed packaging and tool configurations to using `pyproject.toml`.
- Changed certain fields in `ArgumentSpecs`; added a note that this class is internal and offers no stability 
  guarantees.

### Fixed
- Fixed a bug where arguments are not collected in the correct method resolution order (MRO).

## [0.3.1]
### Added
- Support for `snake_case` (underscore style) arguments. (#6)
- A `DeprecationWarning` is now shown when using `Choices` instead of `Literal`.

## [0.3.0]
### Added
- Support for list arguments. (#1)
- `__repr__` method for `Arguments`. (#2)
- Defined arguments are now stored in a special class variable `__arguments__` in the `Arguments` subclass namespace.
  A utility function `argtyped.argument_specs(Args)` is provided to inspect the specifications of the arguments.

### Changed
- Annotations are now parsed on class construction, and the `argparse.ArgumentParser` object is stored for the whole
  lifetime of the class as `__parser__`.
- Exceptions thrown for invalid type annotations are changed from `ValueError`s to `TypeError`s.

### Fixed
- It is now possible to override an argument with default value defined in the base class with a new argument that does
  not have a default. Namely, the following code is now valid (although discouraged):
  ```python
  from argtyped import Arguments, Choices
  
  class BaseArgs(Arguments):
      foo: int = 0
  
  class DerivedArgs(BaseArgs):
      foo: Choices["a", "b"]
  ```
- `Optional[Literal[...]]` is now correctly supported.
- `Optional[Switch]` is now correctly detected (although invalid).

## [0.2.0] - 2020-06-15
### Added
- Literal types: `Literal`. They act mostly the same as `Choices`.

### Changed
- `Arguments` is now pickle-able.

## [0.1] - 2020-02-16
### Added
- The base of everything: the `Argument` class.
- Choice types: `Choices`.
- Custom enum types: `Enum`. They differ from built-in `Enum` in that `auto()` produces the enum name instead of
  numbers.
- Switch types: `Switch`.

[Unreleased]: https://github.com/huzecong/argtyped/compare/v0.4.0...HEAD
[0.4.0]: https://github.com/huzecong/argtyped/compare/v0.3.1...v0.4.0
[0.3.1]: https://github.com/huzecong/argtyped/compare/v0.3.0...v0.3.1
[0.3.0]: https://github.com/huzecong/argtyped/compare/v0.2.0...v0.3.0
[0.2.0]: https://github.com/huzecong/argtyped/compare/v0.1...v0.2.0
[0.1]: https://github.com/huzecong/argtyped/releases/tag/v0.1


argtyped/arguments.py
import argparse
import enum
import functools
import shutil
import sys
from abc import ABCMeta
from collections import OrderedDict
from typing import (
    Any,
    Callable,
    Dict,
    List,
    NamedTuple,
    NoReturn,
    Optional,
    Tuple,
    Type,
    TypeVar,
    Union,
)

from argtyped.custom_types import (
    Switch,
    is_choices,
    is_enum,
    is_list,
    is_optional,
    unwrap_choices,
    unwrap_list,
    unwrap_optional,
)

__all__ = [
    "Arguments",
    "ArgumentKind",
    "ArgumentSpec",
    "argument_specs",
]

T = TypeVar("T")
ConversionFn = Callable[[str], T]


class ArgumentParser(argparse.ArgumentParser):
    r"""A class to override some of ``ArgumentParser``\ 's behaviors."""

    def _get_value(self, action: argparse.Action, arg_string: str) -> Any:
        r"""
        The original ``_get_value`` method catches exceptions in user-defined
        ``type_func``\ s and ignores the error message. Here we don't do that.
        """
        type_func = self._registry_get("type", action.type, action.type)

        try:
            result = type_func(arg_string)
        except (argparse.ArgumentTypeError, TypeError, ValueError) as e:
            message = f"value '{arg_string}', {e.__class__.__name__}: {str(e)}"
            raise argparse.ArgumentError(action, message)

        return result

    def error(self, message: str) -> NoReturn:
        r"""
        The original ``error`` method only prints the usage and force quits. Here we
        print the full help.
        """
        self.print_help(sys.stderr)
        sys.stderr.write(f"{self.prog}: error: {message}\n")
        self.exit(2)

    def add_switch_argument(
        self, name: str, default: bool = False, underscore: bool = False
    ) -> None:
        r"""
        Add a "switch" argument to the parser. A switch argument with name ``"flag"``
        has value ``True`` if the argument ``--flag`` exists, and ``False`` if
        ``--no-flag`` exists.
        """
        assert name.startswith("--")
        name = name[2:]
        var_name = name.replace("-", "_")
        self.add_argument(
            f"--{name}", action="store_true", default=default, dest=var_name
        )
        off_arg_name = f"--no_{name}" if underscore else f"--no-{name}"
        self.add_argument(off_arg_name, action="store_false", dest=var_name)


def _bool_conversion_fn(s: str) -> bool:
    if s.lower() in {"y", "yes", "true", "ok"}:
        return True
    if s.lower() in {"n", "no", "false"}:
        return False
    raise ValueError(f"Invalid value {s!r} for bool argument")


def _optional_wrapper_fn(fn: ConversionFn[T]) -> ConversionFn[Optional[T]]:
    @functools.wraps(fn)
    def wrapped(s: str) -> Optional[T]:
        if s.lower() == "none":
            return None
        return fn(s)

    return wrapped


_TYPE_CONVERSION_FN: Dict[type, ConversionFn[Any]] = {
    bool: _bool_conversion_fn,
}


class ArgumentKind(enum.Enum):
    """
    The kind of argument:

    - ``NORMAL``: A normal argument that takes a single value.
    - ``SWITCH``: A boolean switch argument that takes no values; it is set to True with
      ``--argument`` and False with ``--no-argument``
      (``action="store_true"/"store_false"``).
    - ``SEQUENCE``: A sequential argument that takes multiple values (``nargs="*"``).
    """

    NORMAL = 0
    SWITCH = 1
    SEQUENCE = 2


_NOTHING = object()  # sentinel


class ArgumentSpec(NamedTuple):
    """
    Internal specs of an argument.

    This class is internal -- there's no stability guarantees on its attributes across
    versions.
    """

    name: str
    nullable: bool
    required: bool
    type: type
    kind: ArgumentKind
    choices: Optional[Tuple[Any, ...]] = None
    default: Any = _NOTHING

    positional: bool = False
    # ^ Whether the argument is a positional argument.  If False, it is a keyword(-only)
    #   argument.
    argparse_options: Optional[Dict[str, Any]] = None
    # ^ Additional arguments to pass to `ArgumentParser.add_argument`.  This takes
    #   precedence over `argtyped`'s computed options, e.g. you can set `nargs="+"` for
    #   sequence-type arguments.
    parse: bool = True
    # ^ Whether the argument value should be parsed.  If False, it is the downstream's
    #   responsibility to parse (e.g. in `attrs`).
    underscore: bool = False
    # ^ Argument naming convention:
    #   True for `--snake_case` args, False for `--kebab-case` (default).
    inherited: bool = False
    # ^ True if argument was defined in a base class and is not overridden in the
    #   current class.

    def with_options(self, positional: bool = False, **kwargs: Any) -> "ArgumentSpec":
        """Return a new spec with additional ``argparse`` options."""
        return self._replace(  # pylint: disable=no-member
            positional=positional, argparse_options=kwargs or None
        )

    def with_default(self, value: Optional[Any]) -> "ArgumentSpec":
        """
        Return a new spec with a default value, and perform validation on the value.

        By default, we don't store default values on the spec.  This is because default
        value handling could happen outside ``argtyped``, e.g. for
        :class:`AttrsArguments` it is handled by ``attrs``.
        """

        def value_error(message: str) -> NoReturn:
            raise TypeError(
                f"Argument {self.name!r} has invalid default value {value!r}: {message}"
            )

        if not self.nullable and value is None:
            value_error(
                "Argument not nullable. Change type annotation to 'Optional[...]' "
                "to allow values of None"
            )
        if self.kind == ArgumentKind.SWITCH:
            if not isinstance(value, bool):
                value_error("Switch argument must have a boolean default value")
        if self.kind == ArgumentKind.SEQUENCE:
            if not isinstance(value, list):
                value_error("Default for list argument must be of list type")
            value_seq = value
        else:
            value_seq = [value]
        if self.choices is not None:
            if not all(
                x in self.choices  # pylint: disable=unsupported-membership-test
                or (x is None and self.nullable)
                for x in value_seq
            ):
                value_error("Value must be among valid choices")

        return self._replace(default=value, required=False)  # pylint: disable=no-member


def _generate_argument_spec(
    arg_name: str,
    arg_type: Any,
    has_default: bool,
    *,
    underscore: bool = False,
) -> ArgumentSpec:
    original_type = arg_type

    def type_error(message: str) -> None:
        raise TypeError(
            f"Argument {arg_name!r} has invalid type {original_type!r}: {message}"
        )

    if isinstance(arg_type, str):
        type_error("forward references are not yet supported")

    # On sequence and nullable types:
    # - Nested lists (e.g. `List[List[T]]`) are not supported.
    # - When mixing `List` and `Optional`, the only allowed variant is
    #   `List[Optional[T]]`. Anything else (`Optional[List[T]]`,
    #   `Optional[List[Optional[T]]]`) is invalid.
    sequence = is_list(arg_type)
    if sequence:
        arg_type = unwrap_list(arg_type)
    nullable = is_optional(arg_type)
    if nullable:
        arg_type = unwrap_optional(arg_type)
    if (sequence or nullable) and (arg_type is Switch or is_list(arg_type)):
        type_error(
            f"{'List' if is_list(arg_type) else 'Switch'!r} cannot be "
            f"nested inside {'List' if sequence else 'Optional'!r}",
        )

    if arg_type is Switch:
        return ArgumentSpec(
            name=arg_name,
            kind=ArgumentKind.SWITCH,
            required=False,
            nullable=False,
            type=bool,
            underscore=underscore,
        )

    choices = None
    if is_enum(arg_type) or is_choices(arg_type):
        if is_enum(arg_type):
            value_type = arg_type
            choices = tuple(arg_type)
        else:
            value_type = str
            choices = unwrap_choices(arg_type)
            if any(not isinstance(choice, str) for choice in choices):
                type_error("All choices must be strings")
    else:
        if arg_type not in _TYPE_CONVERSION_FN and not callable(arg_type):
            type_error("Unsupported type")
        value_type = arg_type
    return ArgumentSpec(
        name=arg_name,
        kind=ArgumentKind.SEQUENCE if sequence else ArgumentKind.NORMAL,
        required=not has_default,
        nullable=nullable,
        type=value_type,
        choices=choices,
        underscore=underscore,
    )


def _build_parser(
    arguments: "OrderedDict[str, ArgumentSpec]", cls: type
) -> ArgumentParser:
    """Create the :class:`ArgumentParser` for this :class:`Arguments` class."""
    parser = ArgumentParser()
    for name, spec in arguments.items():
        arg_name = name if spec.underscore else name.replace("_", "-")
        if not spec.positional:
            arg_name = f"--{arg_name}"
        if spec.kind in {ArgumentKind.NORMAL, ArgumentKind.SEQUENCE}:
            arg_type = spec.type
            kwargs: Dict[str, Any] = {}
            if spec.positional:
                if not spec.required:
                    kwargs["nargs"] = "?"
                    kwargs["default"] = spec.default
            else:
                if spec.required:
                    kwargs["required"] = True
                else:
                    kwargs["default"] = spec.default
            if spec.parse:
                conversion_fn = _TYPE_CONVERSION_FN.get(arg_type, arg_type)
                if spec.nullable:
                    conversion_fn = _optional_wrapper_fn(conversion_fn)
                kwargs["type"] = conversion_fn
            if spec.choices is not None:
                if spec.nullable:
                    kwargs["choices"] = spec.choices + (None,)
                else:
                    kwargs["choices"] = spec.choices
                if is_enum(spec.type):
                    # Display only the enum names in help.
                    kwargs["metavar"] = (
                        "{" + ",".join(val.name for val in spec.choices) + "}"
                    )
            if spec.kind == ArgumentKind.SEQUENCE:
                kwargs["nargs"] = "*"
            if spec.argparse_options is not None:
                kwargs.update(spec.argparse_options)
            parser.add_argument(arg_name, **kwargs)
        else:
            assert spec.default is not None
            parser.add_switch_argument(arg_name, spec.default, spec.underscore)

    if cls.__module__ != "__main__":
        # Usually arguments are defined in the same script that is directly
        # run (__main__). If this is not the case, add a note in help message
        # indicating where the arguments are defined.
        parser.epilog = f"Note: Arguments defined in {cls.__module__}.{cls.__name__}"
    return parser


class ArgumentsMeta(ABCMeta):
    r"""
    Metaclass for :class:`Arguments`. The type annotations are parsed and converted into
    an ``argparse.ArgumentParser`` on class creation.
    """
    __parser__: ArgumentParser
    __arguments__: "OrderedDict[str, ArgumentSpec]"

    def __new__(  # type: ignore[misc]
        mcs,
        name: str,
        bases: Tuple[type, ...],
        namespace: Dict[str, Any],
        **kwargs: Any,
    ) -> "Type[Arguments]":
        cls: "Type[Arguments]" = super().__new__(  # type: ignore[assignment]
            mcs, name, bases, namespace
        )

        root = kwargs.get("_root", False)
        if not root and not issubclass(cls, Arguments):
            raise TypeError(f"Type {cls.__name__!r} must inherit from `Arguments`")
        if root:
            cls.__arguments__ = OrderedDict()
            return cls

        arguments: "OrderedDict[str, ArgumentSpec]" = OrderedDict()
        for base in reversed(cls.mro()[1:]):
            # Use reversed order so derived classes can override base annotations.
            if issubclass(base, Arguments):
                for arg_name, spec in argument_specs(base).items():
                    if spec.inherited:
                        # Skip inherited attributes -- they should have been included
                        # already when we processed the base class, which is higher up
                        # in the MRO.
                        continue
                    arguments[arg_name] = spec._replace(inherited=True)

        # Check if there are arguments with default values but without annotations.
        annotations = getattr(cls, "__annotations__", {})
        for key in annotations:
            if key.startswith("_") and not key.startswith("__"):
                raise TypeError(f"Argument {key!r} must not start with an underscore")
        annotations = OrderedDict(
            [(k, v) for k, v in annotations.items() if not k.startswith("__")]
        )
        for key, value in namespace.items():
            if not key.startswith("_") and not callable(value):
                if key not in annotations and key not in arguments:
                    raise TypeError(f"Argument {key!r} does not have type annotation")

        # Check validity of arguments and create specs.
        underscore = kwargs.get("underscore", False)
        for arg_name, arg_type in annotations.items():
            has_default = arg_name in namespace
            spec = _generate_argument_spec(
                arg_name, arg_type, has_default, underscore=underscore
            )
            if has_default:
                spec = spec.with_default(namespace[arg_name])
            elif spec.kind == ArgumentKind.NORMAL and spec.nullable:
                spec = spec.with_default(None)
            arguments[arg_name] = spec

        # The parser will be lazily constructed when the `Arguments` instance is first
        # initialized.
        cls.__arguments__ = arguments
        return cls


class Arguments(metaclass=ArgumentsMeta, _root=True):
    r"""
    A typed version of ``argparse``. It's easier to illustrate using an example:

    .. code-block:: python

        from typing import Optional

        from argtyped import Arguments, Choices, Switch
        from argtyped import Enum, auto

        class LoggingLevels(Enum):
            Debug = auto()
            Info = auto()
            Warning = auto()
            Error = auto()
            Critical = auto()

        class MyArguments(Arguments):
            model_name: str
            hidden_size: int = 512
            activation: Choices['relu', 'tanh', 'sigmoid'] = 'relu'
            logging_level: LoggingLevels = LoggingLevels.Info
            use_dropout: Switch = True
            dropout_prob: Optional[float] = 0.5

        args = Arguments()

    This is equivalent to the following code with Python built-in ``argparse``:

    .. code-block:: python

        import argparse

        parser = argparse.ArgumentParser()
        parser.add_argument("--model-name", type=str, required=True)
        parser.add_argument("--hidden-size", type=int, default=512)
        parser.add_argument(
            "--activation", choices=["relu", "tanh", "sigmoid"], default="relu"
        )
        parser.add_argument(
            "--logging-level", choices=ghcc.logging.get_levels(), default="info"
        )
        parser.add_argument(
            "--use-dropout", action="store_true", dest="use_dropout", default=True
        )
        parser.add_argument(
            "--no-use-dropout", action="store_false", dest="use_dropout"
        )
        parser.add_argument(
            "--dropout-prob",
            type=lambda s: None if s.lower() == 'none' else float(s),
            default=0.5
        )

        args = parser.parse_args()

    Suppose the following arguments are provided:

    .. code-block:: bash

        python main.py \
            --model-name LSTM \
            --activation sigmoid \
            --logging-level debug \
            --no-use-dropout \
            --dropout-prob none

    the parsed arguments will be:

    .. code-block:: bash

        Namespace(model_name="LSTM", hidden_size=512, activation="sigmoid",
                  logging_level="debug", use_dropout=False, dropout_prob=None)

    :class:`Arguments` provides the following features:

    - More concise and intuitive syntax over ``argparse``, less boilerplate code.
    - Arguments take the form of type-annotated class attributes, allowing IDEs to
      provide autocompletion.
    - Drop-in replacement for ``argparse``, since internally ``argparse`` is used.

    **Note:** Advanced features such as subparsers, groups, argument lists, custom
    actions are not supported.
    """

    def __init__(self, args: Optional[List[str]] = None):
        cls = self.__class__
        if not hasattr(cls, "__parser__"):
            parser = _build_parser(cls.__arguments__, cls)

            cls.__parser__ = parser
        namespace = cls.__parser__.parse_args(args)
        for arg_name in argument_specs(cls):
            setattr(self, arg_name, getattr(namespace, arg_name))

    def to_dict(self) -> "OrderedDict[str, Any]":
        r"""
        Convert the set of arguments to a dictionary.

        :return: An ``OrderedDict`` mapping argument names to values.
        """
        return OrderedDict(
            [(key, getattr(self, key)) for key in argument_specs(self.__class__).keys()]
        )

    def to_string(
        self, width: Optional[int] = None, max_width: Optional[int] = None
    ) -> str:
        r"""
        Represent the arguments as a table.

        :param width: Width of the printed table. Defaults to ``None``, which fits the
            table to its contents. An exception is raised when the table cannot be drawn
            with the given width.
        :param max_width: Maximum width of the printed table. Defaults to ``None``,
            meaning no limits. Must be ``None`` if :arg:`width` is not ``None``.
        """
        if width is not None and max_width is not None:
            raise ValueError("`max_width` must be None when `width` is specified")

        k_col = "Arguments"
        v_col = "Values"
        arg_reprs = {k: repr(v) for k, v in self.to_dict().items()}
        max_key = max(len(k_col), max(len(k) for k in arg_reprs.keys()))
        max_val = max(len(v_col), max(len(v) for v in arg_reprs.values()))
        margin_col = 7  # table frame & spaces
        if width is not None:
            max_val = width - max_key - margin_col
        elif max_width is not None:
            max_val = min(max_val, max_width - max_key - margin_col)
        if max_val < len(v_col):
            raise ValueError("Table cannot be drawn under the width constraints")

        def get_row(k: str, v: str) -> str:
            if len(v) > max_val:
                v = v[: ((max_val - 5) // 2)] + " ... " + v[-((max_val - 4) // 2) :]
                assert len(v) == max_val
            return f"║ {k.ljust(max_key)} │ {v.ljust(max_val)} ║\n"

        s = repr(self.__class__) + "\n"
        s += f"╔═{'═' * max_key}═╤═{'═' * max_val}═╗\n"
        s += get_row(k_col, v_col)
        s += f"╠═{'═' * max_key}═╪═{'═' * max_val}═╣\n"
        for k, v in arg_reprs.items():
            s += get_row(k, v)
        s += f"╚═{'═' * max_key}═╧═{'═' * max_val}═╝\n"
        return s

    def __repr__(self) -> str:
        columns, _ = shutil.get_terminal_size()
        return self.to_string(max_width=columns)


def argument_specs(
    args_class: Union[Arguments, Type[Arguments]]
) -> "OrderedDict[str, ArgumentSpec]":
    r"""
    Return a dictionary mapping argument names to their specs (:class:`ArgumentSpec`
    objects).
    """
    if isinstance(args_class, Arguments):
        return args_class.__class__.__arguments__
    return args_class.__arguments__
